django/views/generic/detail.py

```diff
@@ -100,6 +100,11 @@ def get_context_data(self, **kwargs):
         Insert the single object into the context dict.
         """
         context = {}
+        # when self.object value is None, we should NOT run get_object()
+        # method; this value means object will be created by ModelForm (in
+        # CreateView)
+        if not hasattr(self, "object"):
```

Comment:
I understand reservations about this change and I thought a lot about pros and cons of it, especially in context of !4595.

Hopefully @freakboy3742 can look at the following comment and see if I got it right, since he knows about the design decisions of CBGVs.

In current CBGVs, things are generally passed around as arguments as it allows more flexibility. The exceptions are self.object and self.object_list. These are set in get() and post() and are used in subsequent methods, and are available for use (and are often used) in all methods overridden by users. The reason for this is that they are central to the respective GCBVs and may or may not be useful in many methods, and it would be tedious to pass them around as args.

Another possible reason is that in case of self.object, a db hit is required, so you want to make sure users don't get it multiple inadvertantly. It makes it easier to avoid this if it is always set in get() or post() (which would not be both used in the same request).

So at first look current design makes a lot of sense.

However, when you actually try to do common-case customizations of CBGVs, as I've done in the PR I mentioned, it becomes problematic. For example, it creates a lot more legwork to mix different CBGVs.

Officially, they're not designed to be mixed and you either shouldn't or be prepared to deal with resulting complexity. In practice it's often possible to mix CBGVs and it's often easy to do.

My goal was to look at a fair number of customizations and see if there are changes that can make them much easier without drastic changes to current design and keeping backwards compatibility.

The get_context_data() method is different from other methods in that it accumulates context from all parent classes which makes it already very convenient to use when mixing different CBGVs. Therefore it would be really nice if these methods had a self-contained quality where you could run any number of them and they would do whatever is necessary to add their respective contexts.

Therefore, you could use a single stock get() method (or a custom one) and an arbitrary number of get_context_data() methods and this PR would make things easier in two ways:

1. Often you won't need to override get() or post() (or especially if you need to override both in the same view), or, if overriding, you can make it simpler by letting get_context_data() do their own setup.

2. When making changes you can modify all of get(), post(), and get_context_data() without having to keep track of their dependencies.

The only cost I see is that it adds a bit of magicality, and can make it a bit harder to suppress creation of self.object in Detail or ModelForm views, but it seems a much rarer use case than the kind of customizations I've used as examples in !4595.